### PR TITLE
feat: Add option retry_timeout that like nginx's proxy_next_upstream_timeout

### PR DESCRIPTION
**Pull request by chzhuo**
Add option `retry_timeout` to Limits the time during which a request can be passed to next upstream.

### What this PR does / why we need it:
option `retry_timeout` like `nginx`'s directive  [proxy_next_upstream_timeout](http://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_next_upstream_timeout)

### Pre-submission checklist:

* [x] Did you explain what problem does this PR solve? Or what new features have been added?
* [x] Have you added corresponding test cases?
* [x] Have you modified the corresponding document?
* [x] Is this PR backward compatible? **If it is not backward compatible, please discuss on the [mailing list](https://github.com/apache/apisix/tree/master#community) first**

### CODE REVIEW

File: t/admin/upstream.t [hunk 1]
```diff
@@ -1978,3 +1978,133 @@ GET /t
 [delete] code: 200 message: passed
 --- no_error_log
 [error]
+
+
+
+=== TEST 57: retry_timeout is -1 (INVALID)
```
Comment (spacewander) [inline t/admin/upstream.t:1984]: Please add test in t/admin/upstream3.t

File: t/lib/server.lua [hunk 2]
```diff
@@ -417,5 +417,13 @@ function _M.server_error()
     error("500 Internal Server Error")
 end
 
+function _M.retry_error()
```
Comment (spacewander) [inline t/lib/server.lua:420]: What about using `_M.mysleep`?

**Review by Yiyiyimu — APPROVED** [review]
LGTM

File: docs/zh/latest/admin-api.md [hunk 3]
```diff
@@ -551,6 +551,7 @@ APISIX 的 Upstream 除了基本的负载均衡算法选择外，还支持对上
...
 | retries        | 可选                               | 整型           | 使用底层的 Nginx 重试机制将请求传递给下一个上游，默认启用重试且次数为后端可用的 node 数量。如果指定了具体重试次数，它将覆盖默认值。`0` 代表不启用重试机制。                                                                                                                                                                                                 |                                                  |
+| retry_timeout  | 可选                               | 整型           | 限制请求传递给下一个上游的时间。`0` 代表不启用重试机制。                                                                                                                                                                                                 |                                                  |
```
Comment (tokers) [inline docs/zh/latest/admin-api.md:554]: IMHO the description is not easy to understand.

Comment (tokers) [reply docs/zh/latest/admin-api.md:554]: It's more like a timeout setting which limits the whole time that all retried requests cost.

File: t/admin/upstream3.t [hunk 4]
```diff
@@ -61,3 +61,35 @@ __DATA__
...
+            end
+            ngx.print(body)
+        }
+    }
+--- request
+GET /t
+--- error_code: 400
+--- response_body
+{"error_msg":"invalid configuration: property \"retry_timeout\" validation failed: expected -1 to be greater than 0"}
+--- no_error_log
```
Comment (spacewander) [inline t/admin/upstream3.t:94]: The no_error_log already set in https://github.com/apache/apisix/blob/42ab6149272920bbab66985b90e2754a7c6cb747/t/admin/upstream3.t#L32

**Review by spacewander — CHANGES_REQUESTED** [review]

File: docs/zh/latest/admin-api.md (hunk 3 above)
Comment (chzhuo) [reply docs/zh/latest/admin-api.md:554]: 限制是否继续重试的时间，若之前的请求和重试请求花费太多时间就不再继续重试。Limit the amount of time that retries can be continued, and do not continue retries if the previous request and retry requests have taken too long

File: t/admin/upstream3.t [hunk 5]
```diff
@@ -91,5 +91,3 @@ GET /t
 --- error_code: 400
 --- response_body
 {"error_msg":"invalid configuration: property \"retry_timeout\" validation failed: expected -1 to be greater than 0"}
---- no_error_log
```
Comment (tokers) [inline t/admin/upstream3.t]: Why remove this?

Comment (chzhuo) [reply t/admin/upstream3.t]: https://github.com/apache/apisix/blob/42ab6149272920bbab66985b90e2754a7c6cb747/t/admin/upstream3.t#L32-L34
In file `upstream3.t` has inserted this block when no error

**Review by spacewander — APPROVED** [review]

**Review by tokers — APPROVED** [review]

**Comment by tokers** [comment]
@chzhuo Please resolve the conflicts.

**Pull request merged**